cmd/vicadmin/server.go

```diff
@@ -112,39 +127,208 @@ func (s *server) listenPort() int {
+		if c == nil { // no cookie, so redirect to login
+			log.Errorf("No authentication token: %+v", websession.Values)
+			http.Redirect(w, r, loginPagePath, http.StatusTemporaryRedirect)
+			return
+		}
+
+		// here we have a cookie, but we need to make sure it's not expired:
+		// parse the cookie creation time
+		created, err := time.Parse(time.RFC3339, c.(string))
+		if err != nil {
+			// we pulled this value out of a cookie, so if it doesn't parse, it might've been tampered with
+			// though the cookie's encrypted so that would destroy the whole cookie..
+			// Handling the error in any case:
+			log.Errorf("Couldn't parse time out of retrieved cookie due to error %s", err.Error())
+			http.Error(w, genericErrorMessage, http.StatusInternalServerError)
+			return
+		}
+
+		// cookie exists but is expired
+		if time.Since(created) > sessionExpiration {
+			http.Redirect(w, r, loginPagePath+"?expired", http.StatusTemporaryRedirect)
+			return
+		}
+
+		// if the date on the cookie was valid, then the user is authenticated
+		handler(w, r)
+	}
+	s.mux.Handle(link, gorillacontext.ClearHandler(http.HandlerFunc(authHandler)))
+}
+
+// renders the page for login and handles authorization requests
+func (s *server) loginPage(res http.ResponseWriter, req *http.Request) {
+	defer trace.End(trace.Begin(""))
+	ctx := context.Background()
+	if req.Method == "POST" {
+		// take the form data and use it to try to authenticate with vsphere
+
+		// create a userconfig
+		userconfig := session.Config{
+			Insecure:       false,
+			Thumbprint:     rootConfig.Thumbprint,
+			Keepalive:      rootConfig.Keepalive,
+			ClusterPath:    rootConfig.ClusterPath,
+			DatacenterPath: rootConfig.DatacenterPath,
+			DatastorePath:  rootConfig.DatastorePath,
+			HostPath:       rootConfig.Config.HostPath,
+			PoolPath:       rootConfig.PoolPath,
+		}
+		user := url.UserPassword(req.FormValue("username"), req.FormValue("password"))
+		serviceURL, err := soap.ParseURL(rootConfig.Service)
+		if err != nil {
+			// this could happen for a number of reasons but most likely for a plain ol' auth failure
+			log.Errorf("vSphere service URL was not a valid format; parsing returned error: %s", err)
+			http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+			return
+		}
+
+		serviceURL.User = user
+		userconfig.Service = serviceURL.String()
+
+		// check login
+		usersession, err := vSphereSessionGet(&userconfig)
+		if err != nil || usersession == nil {
+			// something went wrong or we could not authenticate
+			log.Warnf("User %s from %s failed to authenticated at %s", user, req.RemoteAddr, time.Now())
+			http.Error(res, "Authentication failed due to incorrect credential(s)", 400)
+			return
+		}
+
+		// successful login above; user is authenticated
+		// log out, disregard errors
+		usersession.Client.Logout(context.Background())
+
+		// save user config locally
+		usersess := s.uss.Add(req.FormValue("username"), &userconfig)
```

Comment:
Yep just checked and `websession` has a `string` `ID` so I'll use that as the unique identifier instead of the username -- I feel really silly for using the username actually hahaha